### ISSUE TITLE
Make Event Hubs library mockable

### ISSUE THREAD

**Issue by cwe1ss** [Azure/azure-event-hubs-dotnet#24]
Right now, it's really hard to test code using this library because many important concepts are marked as internal.
One example is that you can't even create an `EventData` object with system properties because the [setter is internal](https://github.com/Azure/azure-event-hubs-dotnet/blob/master/src/Microsoft.Azure.EventHubs/EventData.cs#L80). 

It would be great if you could follow the example of ASP.NET Core (and other new MS open source projects) and use more interfaces, protected setters, etc. and an "Internal" folder/namespace instead of internal types. All types in that folder would not be backed by the typical backwards compatibility and usage would be at everyone's own risk.

**Comment by serkantkaraca** [comment]
Thanks Christian for these invaluable feedback.  We will take them into our consideration as we continue improving our client.

**Comment by SeanFeldman** [comment]
Suggestion: rather than turning setters into public, perhaps have a factory method that forces to pass in all the parameters to initialize a new `EventData`? Same with `SystemPropertiesCollection`.

**Comment by jtaubensee** [comment]
Reiterating my comment from PR #120. 

> Our team is planning on properly abstracting the components, and using interfaces to allow easy mocking for test purposes. Please stay tuned for that...

**Comment by kamil-mrzyglod** [comment]
@jtaubensee Cool, good to know!

**Comment by jtaubensee** [comment]
Comment from @enihcam on #125:
> EventHubClient cannot be mocked:
>  * It has no default constructor
>  * SendAsync is not virtual, so it cannot be overridden.

**Comment by erick-thompson** [comment]
Did anything ever happen with this? Not having this makes it hard to test code that interacts with EventHub.

**Comment by punkcoder** [comment]
Is there a possibility of getting a legitimate fix for this item?  I understand that there is a desire for the code to be clean put there are quick fixes for this issue that would allow people to test their implementations of the code more easily... Mostly I'm asking since it looks like this 'refactor' has been open for a year and a half?  It appears that others have attempted to fix this and had their PR's rejected.

**Comment by serkantkaraca** [comment]
I don't really recall any rejected PRs. Anyone is welcome to contribute to this library. Making library Mockable is one of our goals but this is being punted due to limited developer resources.

**Comment by NickDarvey** [comment]
@jtaubensee how's this going?

**Comment by jtaubensee** [comment]
@NickDarvey - Sorry for the delay, but I am no longer with the Messaging team. Perhaps someone else from the team has further thoughts?

**Comment by idg10** [comment]
There has been one rejected PR: #120 - on the grounds that adding members specifically for test purposes is bad.

I might be able to get some time to put together a PR for this, but I don't want to do all the work and then find that it gets rejected because the team doesn't like it much. So I want to understand how you'd like this to look.

Are you envisaging a solution in which all the concrete types currently consumed by the client have corresponding interface types? E.g., something like this (and similar things for all public types):

```csharp
public interface IEventData : IDisposable
{
    ArraySegment<byte> Body { get; }
    IDictionary<string, object> Properties { get; }
    SystemPropertiesCollection SystemProperties { get; }
}
```

(This would entail making the `SystemPropertiesCollection` ctor public. Alternatively we could introduce an `ISystemPropertiesCollection` but my view is that adding interfaces for types that are essentially plain old data is inappropriate. It's the right thing to do for types that have behaviour, but if since `SystemPropertiesCollection` is really just a bag of properties (based on `Dictionary<string, object>`) with some helpers for reading well-known properties, it seems like overkill to define a corresponding interface - like defining an `IString` for properties of type `string`.)

**Comment by gambrose** [comment]
> why exactly does `EventData` have its existing public constructors? If you're going to have public ctors at all, it would make sense for them to be able to create a fully-constructed instance. (However, the right answer might be: these should never have been public because in fact there is no way to create a fully-constructed instance with them

@idg10  `EventData` is used to publish to EventHub. The public constructors are required to create `EventData` objects so that they can be sent to the `PartitionSender`.

**Comment by gambrose** [comment]
> I might be able to get some time to put together a PR for this, but I don't want to do all the work and then find that it gets rejected because the team doesn't like it much. So I want to understand how you'd like this to look.

@serkantkaraca Do you have any comments on what changes would be most likely to be accepted?
I have a similar feeling to @idg10, I am reluctant to do any work on this if it is going to be rejected or ignored. With this issue approaching it's second birthday and still no direction from Azure. It does not inspire confidence.

I would be keen for this issue to progress. At the moment I am using reflection to new up an `EventData` with `SystemProperties`.

**Comment by erick-thompson** [comment]
It would be very nice to have some movement on this. It makes writing tests far harder than it should be.

**Comment by serkantkaraca** [comment]
I am preparing a PR to address mock-ability issues. Can you please add where you are failing to mock?

I will start with the most crucial one; EventData
- SystemProperties
- LastSequenceNumber
- LastEnqueuedOffset
- LastEnqueuedTime
- RetrievalTime

**Comment by gambrose** [comment]
@serkantkaraca That list covers my use case for now.

**Comment by serkantkaraca** [comment]
Here is the PR - https://github.com/Azure/azure-event-hubs-dotnet/pull/332

Started with public SystemProperties. I cannot see any value yet at exposing other properties of EventData publicly. Any thoughts regarding the rest of the properties in the list?

**Comment by davidrevoledo** [comment]
Hi guys I've other concern as well,

To be able to run unit tests in this library azure infrastructure is required, it'd be handy run those test without hit real azure services at least for all.

I'm not meaning about external projects that need to mock the library but internally, this also would help to do the library more mockable  as well :) 

What do you think about it ?

**Comment by Plasma** [comment]
I've also had to write an adapter class for ``PartitionContext`` since it has no public constructor.

**Comment by NatanaelLemos** [comment]
Hi @Plasma , do you have any update on this PartitionContext change?
Also, once this is done, how do you suggest to mock the Lease object inside of PartitionContext (since its setter is internal as well).

**Comment by Plasma** [comment]
Hey @NatanaelLemos we ended up creating some additional methods to consume data that takes no PartitionContext (and our internal code just sees it as optional, now).

For reference, we made this class for our own use around EventData:

```csharp
	/// <summary>
	/// Wrapper around Azure EventHub's EventData class to support unit testing
	/// </summary>
	public class EventDataAdapter
	{
		public EventDataAdapter(byte[] body, DateTime enqueuedTimeUtc = default(DateTime), long sequenceNumber = 0, string offset = null, IDictionary<string, object> properties = null, Microsoft.Azure.EventHubs.EventData eventData = null)
		{
			Body = new ArraySegment<byte>(body);
			EnqueuedTimeUtc = enqueuedTimeUtc;
			SequenceNumber = sequenceNumber;
			Offset = offset;
			Properties = properties ?? new Dictionary<string, object>();
			EventData = eventData;
		}

		public EventDataAdapter(Microsoft.Azure.EventHubs.EventData eventData)
		{
			Body = eventData.Body;
			EnqueuedTimeUtc = eventData.SystemProperties.EnqueuedTimeUtc;
			SequenceNumber = eventData.SystemProperties.SequenceNumber;
			Offset = eventData.SystemProperties.Offset;
			Properties = eventData.Properties;
			EventData = eventData;
		}

		/// <summary>
		/// Returns the underlying EventData if we are operating with a real Azure EventData context
		/// </summary>
		internal Microsoft.Azure.EventHubs.EventData EventData { get; }

		/// <summary>Gets the logical sequence number of the event within the partition stream of the Event Hub.</summary>
		public long SequenceNumber { get; }

		/// <summary>
		/// Gets the offset of the data relative to the Event Hub partition stream. The offset is a marker or identifier for an event within the Event Hubs stream. The identifier is unique within a partition of the Event Hubs stream.
		/// </summary>
		public string Offset { get; }

		/// <summary>
		/// Get the actual Payload/Data wrapped by EventData.
		/// This is intended to be used after receiving EventData using <see cref="T:Microsoft.Azure.EventHubs.PartitionReceiver" />.
		/// </summary>
		public ArraySegment<byte> Body { get; }

		/// <summary>Gets or sets the date and time of the sent time in UTC.</summary>
		/// <value>The enqueue time in UTC. This value represents the actual time of enqueuing the message.</value>
		public DateTime EnqueuedTimeUtc { get; }

		/// <summary>Gets the user properties of the event data that the user explicitly added during send operations.</summary>
		/// <returns>The user properties of the event data.</returns>
		public IDictionary<string, object> Properties { get; }
	}
```

**Comment by NatanaelLemos** [comment]
@Plasma  thank you for your quick reply.
And how do you test your EventProcessor class? In my case, I need to access some of PartitionContext's values.
@serkantkaraca , do you have any feedback on this?

**Comment by serkantkaraca** [comment]
Public PartitionContext https://github.com/Azure/azure-event-hubs-dotnet/pull/373

**Comment by alexanderbikk** [comment]
Hello @NatanaelLemos 
As far as I understand  this improve #373  will be available in version 3.0.1 since it already in master. Am I right?

Thanks

**Comment by NatanaelLemos** [comment]
Hi @alexanderbikk 
I guess so, apparently it's merged into dev branch. But I think you should ask someone from the EventHubs team about it.

**Comment by serkantkaraca** [comment]
Track 2 client will be addressing the issue. We will continue addressing high priority test-ability issues only for this client.

**Comment by dserranoaPsl** [comment]
Hi @Plasma ! How can i Use that EventDataAdapter? I am trying to create an ProcessEventArgs object but before i need to create the EventData object. I am a little bit lost here. Can you give me a hand? Thanks in advance.

**Comment by Plasma** [comment]
Hi @dserranoaPsl `EventDataAdapter` class above is very simple, just a basic class that looks like (but does not inherit from) the `EventData` class. We used this in our code base instead of passing an `EventData` object when possible, so we could use a manually constructed version for unit test purposes.

It seems like since this issue was made, there have been some improvements to `EventData` class and I'd see if you can now create one directly instead of needing `EventDataAdapter`. I'm not familiar with `ProcessEventArgs` object, but perhaps you can follow a similar pattern for my `EventDataAdapter` class (make your own `ProcessEventArgsAdapter` class that looks similar to `ProcessEventArgs` official library class) and use that in your code, instead of the library class that may not be mockable.

**Comment by drdamour** [comment]
was EventHubClient  made mockable with these changes?

**Comment by js8080** [comment]
It looks like #373 done by @serkantkaraca in August 2019 was never actually merged into master. I see the current version of the SDK (5.6.2) still has PartitionContext with a protected internal constructor:

```csharp
namespace Azure.Messaging.EventHubs.Consumer
{
    //
    // Summary:
    //     Represents an Event Hub partition and its relative state, as scoped to an associated
    //     operation performed against it.
    public class PartitionContext
    {
        //
        // Summary:
        //     Initializes a new instance of the Azure.Messaging.EventHubs.Consumer.PartitionContext
        //     class.
        //
        // Parameters:
        //   partitionId:
        //     The identifier of the Event Hub partition this context is associated with.
        protected internal PartitionContext(string partitionId);
```

What's the deal with this? Is it being changed or not? This SDK is difficult to mock which surprises me given that Microsoft has done a good job in other areas with this.

**Comment by EngSayed** [comment]
New Azure SDK uses Azure.Messaging.EventHubs which again has no setter for SystemProperties so it is again not mockable. Is there a solution for that?
